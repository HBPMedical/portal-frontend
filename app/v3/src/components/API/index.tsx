export { default as APICore } from "./Core";
export { default as APIExperiment } from "./Experiment";
export { default as APIMining } from "./Mining";
export { default as APIModel } from "./Model";
