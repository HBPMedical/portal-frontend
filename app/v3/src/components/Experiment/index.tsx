export { default as Highchart } from './Highchart'
export { default as Plotly } from './Plotly'