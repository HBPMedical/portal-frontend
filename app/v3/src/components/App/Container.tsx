import * as React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import request from 'request-promise-native';
import { Provider, Subscribe } from 'unstated';

// import UNSTATED from 'unstated-debug';
import { APICore, APIExperiment, APIMining, APIModel, webURL } from '../API'; // as interfaces
import { Pathology } from '../API/Core';
import config from '../API/RequestHeaders';
import App from '../App/App';

// UNSTATED.logStateChanges = process.env.NODE_ENV === "development";

interface State {
  appConfig: any;
}

class AppContainer extends React.Component<any, State> {
  public state: State = { appConfig: {} };
  private apiExperiment = new APIExperiment(config);
  private apiModel = new APIModel(config);
  private apiCore = new APICore(config);
  private apiMining = new APIMining(config);

  private intervalId: any; // FIXME: NodeJS.Timer | undefined;

  public async componentDidMount() {
    this.intervalId = setInterval(() => this.apiExperiment.all(), 10 * 1000);

    // Conf written by dockerize
    const json = await request.get(`${webURL}/scripts/app/config.json`);
    try {
      const appConfig = JSON.parse(json);
      this.setState({ appConfig });
    } catch (e) {
      const appConfig = {
        instanceName: 'MIP LOCAL',
        mode: 'local',
        theme: 'mip',
        version: 'alpha'
      };
      this.setState({ appConfig });
      console.log("Couldn't read config");
    }

    console.log(this.state.appConfig.mode);

    return await Promise.all([
      this.apiExperiment.all(),
      this.apiCore.datasets(),
      this.apiCore.setPathology(Pathology.DEG),
      this.apiCore.algorithms(this.state.appConfig.mode === 'local'),
      this.apiCore.stats(),
      this.apiModel.all()
    ]);
  }

  public componentWillUnmount() {
    clearInterval(this.intervalId);
  }

  public render() {
    return (
      <Router>
        <Provider
          inject={[
            this.apiExperiment,
            this.apiCore,
            this.apiModel,
            this.apiMining
          ]}>
          <Subscribe to={[APIExperiment, APICore, APIModel, APIMining]}>
            {(
              apiExperiment: APIExperiment,
              apiCore: APICore,
              apiModel: APIModel,
              apiMining: APIMining
            ) => {
              return (
                <App
                  appConfig={this.state.appConfig}
                  apiExperiment={apiExperiment}
                  apiCore={apiCore}
                  apiModel={apiModel}
                  apiMining={apiMining}
                />
              );
            }}
          </Subscribe>
        </Provider>
      </Router>
    );
  }
}

export default AppContainer;
