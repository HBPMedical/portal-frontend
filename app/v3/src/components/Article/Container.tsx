import React, { useEffect, useState } from 'react';
import { Panel } from 'react-bootstrap';
import { RouteComponentProps } from 'react-router-dom';
import { draftStateToHTML } from 'react-wysiwyg-typescript';
import styled from 'styled-components';

import { APICore } from '../API';
import { Article } from '../API/Core';
import Edit from './Edit';

export enum Mode {
  default,
  editing,
  creating
}

const Layout = styled.div`
  padding: 0 48px 0px 48px;
  display: flex;
`;

const PanelLayout = styled(Panel)`
  margin-right: 8px;
`;

const PanelTitle = styled(Panel.Title)`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-right: 8px;
`;

const Submit = styled.div`
  text-align: right;
`;

const initialFormState: Article = { slug: '', title: '' };

interface Params {
  slug: string;
}
interface Props extends RouteComponentProps<Params> {
  apiCore: APICore;
}

export default ({ apiCore, ...props }: Props) => {
  const [currentArticle, setCurrentArticle] = useState<Article>(
    initialFormState
  );
  const [editing, setEditing] = useState(Mode.default);
  const slug = props.match.params.slug;

  useEffect(() => {
    const articles = apiCore.state && apiCore.state.articles;
    
    if (!articles) {
      return;
    }

    if (slug !== undefined) {
      const nextArticle = articles.find(a => a.slug === slug);
      if (nextArticle) {
        setCurrentArticle(nextArticle);
      }
    } else {
      setCurrentArticle(articles[0]);
    }
  }, [slug, apiCore.state]);

  const handleSelectArticle = (article: Article) => {
    setEditing(Mode.default);
    setCurrentArticle(article);
  };

  const handleSaveArticle = (id: string, article: Article) => {
    editing === Mode.editing
      ? updateArticle(article.slug, article)
      : createArticle(article);
  };

  const updateArticle = (id: string, article: Article) => {
    apiCore.updateArticle(id, article).then(() => {
      apiCore.articles();
      setEditing(Mode.default);
    });
  };

  const createArticle = (article: Article) => {
    apiCore.createArticle(article).then(() => {
      apiCore.articles();
      setEditing(Mode.default);
    });
  };

  return (
    <Layout>
      <PanelLayout style={{ flex: 0.3 }}>
        <PanelTitle>
          <h3>My articles</h3>
          <button
            className='btn-primary btn' // tslint:disable-next-line
            onClick={() => {
              setCurrentArticle(initialFormState);
              setEditing(Mode.creating);
            }}>
            New article
          </button>
        </PanelTitle>
        <Panel.Body>
          {apiCore.state &&
            apiCore.state.articles &&
            apiCore.state.articles.map(a => (
              <p key={a.slug}>
                <a
                  // tslint:disable-next-line jsx-no-lambda
                  onClick={() => handleSelectArticle(a)}>
                  {a.title}
                </a>
              </p>
            ))}
        </Panel.Body>
      </PanelLayout>
      <PanelLayout style={{ flex: 0.6 }}>
        <Panel.Body>
          {editing ? (
            <Edit
              currentArticle={currentArticle}
              handleSaveArticle={handleSaveArticle}
              setEditing={setEditing}
            />
          ) : (
            <>
              <Submit>
                <button
                  className='btn-primary btn'
                  // tslint:disable-next-line
                  onClick={() => setEditing(Mode.editing)}>
                  Edit
                </button>
              </Submit>
              <h1>{currentArticle && currentArticle.title}</h1>
              <h3>{currentArticle && currentArticle.abstract}</h3>
              <div>
                {currentArticle &&
                  currentArticle.content &&
                  draftStateToHTML(currentArticle.content)}
              </div>
            </>
          )}
        </Panel.Body>
      </PanelLayout>
    </Layout>
  );
};
