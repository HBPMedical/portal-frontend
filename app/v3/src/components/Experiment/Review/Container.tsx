import queryString from 'query-string';
import * as React from 'react';
import { Panel } from 'react-bootstrap';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import { MIP } from '../../../types';
import { APICore, APIMining, APIModel } from '../../API';
import { IAlert } from '../../UI/Alert';
import Model from '../../UI/Model';
import Validation from '../../UI/Validation';
import Content from './Content';
import Filter from './Filter';
import ExperimentReviewHeader from './Header';

import './Review.css';
interface IProps extends RouteComponentProps<any> {
  apiModel: APIModel;
  apiCore: APICore;
  apiMining: APIMining;
}
interface IState {
  alert?: IAlert;
  loadingSummary?: boolean;
  query?: MIP.API.IQuery;
  mining?: any;
}

class Container extends React.Component<IProps, IState> {
  public state: IState = {};

  public async componentDidMount() {
    const qs = queryString.parse(this.props.location.search);

    if (qs.execute) {
      const variables = [{ code: qs.variable as string }];
      const coVariables =
        qs.covariable && (qs.covariable as string) !== ''
          ? (qs.covariable as string).split(',').map(v => ({
              code: v
            }))
          : undefined;
      const groupings =
        qs.grouping && (qs.grouping as string) !== ''
          ? (qs.grouping as string).split(',').map(v => ({
              code: v
            }))
          : undefined;
      const trainingDatasets =
        qs.trainingDatasets && (qs.trainingDatasets as string) !== ''
          ? (qs.trainingDatasets as string).split(',').map(v => ({
              code: v
            }))
          : undefined;
      const filterQuery =
        (qs.filterQuery && decodeURI(qs.filterQuery as string)) || '';
      const filtersFromParams =
        qs.filter && (qs.filter as string) !== ''
          ? (qs.filter as string).split(',').map(v => ({
              code: v
            }))
          : undefined;

      const query: MIP.Internal.IQuery = {
        coVariables,
        filters: filterQuery,
        filtersFromParams,
        groupings,
        trainingDatasets,
        variables
      };

      const { apiModel } = this.props;
      await apiModel.setMock(query);
      await this.setState({ query });
      this.createMining({ query });
    } else {
      const params = this.urlParams(this.props);
      const slug = params && params.slug;
      if (slug) {
        await this.loadModel({ slug });
        const query = this.state.query;
        if (query) {
          this.createMining({ query });
        }
      }
    }
  }

  public async componentWillReceiveProps(nextProps: IProps, prevState: IState) {
    const params = this.urlParams(nextProps);
    const slug = params && params.slug;
    const prevParams = this.urlParams(this.props);
    const prevSlug = prevParams && prevParams.slug;

    if (slug && prevSlug !== slug) {
      await this.loadModel({ slug });
      const query = this.state.query;
      if (query) {
        this.props.apiMining.clear();
        this.createMining({ query });
      }
    }
  }

  public render() {
    const { apiCore, apiModel, apiMining } = this.props;
    const { query } = this.state;
    const variables = apiCore.state.variables;

    // FIXME: move to Filter, refactor in a pure way
    let fields: any[] = [];
    const buildFilter = (code: string) => {
      if (!variables) {
        return;
      }

      const originalVar = variables.find(
        (variable: MIP.API.IVariableEntity) => variable.code === code
      );

      const output: any = originalVar
        ? {
            id: code,
            label: originalVar.label,
            name: code
          }
        : {};

      if (originalVar && originalVar.enumerations) {
        output.values = originalVar.enumerations.map((c: any) => ({
          [c.code]: c.label
        }));
        output.input = 'select';
        output.operators = ['equal', 'not_equal', 'in', 'not_in'];
      }

      const type = originalVar && originalVar.type;
      if (type === 'real') {
        output.type = 'double';
        output.input = 'number';
        output.operators = [
          'equal',
          'not_equal',
          'less',
          'greater',
          'between',
          'not_between'
        ];
      }

      if (type === 'integer') {
        output.type = 'integer';
        output.input = 'number';
        output.operators = [
          'equal',
          'not_equal',
          'less',
          'greater',
          'between',
          'not_between'
        ];
      }

      return output;
    };

    const keys = ['variables', 'coVariables', 'groupings', 'filtersFromParams'];
    const allVariables: string[] = [];
    if (query) {
      keys.forEach((key: string) => {
        const rows = query[key];
        if (rows) {
          rows.forEach((v: any) => {
            allVariables.push(v.code);
          });
        }
      });
    }

    // backward compatibility
    allVariables.push('subjectageyears');
    allVariables.push('gender');

    // add filter variables
    const extractVariablesFromFilter = (filter: any) =>
      filter.rules.map((r: any) => {
        if (r.rules) {
          extractVariablesFromFilter(r);
        }
        if (r.id) { allVariables.push(r.id); }
      });

    if (query && query.filters) {
      extractVariablesFromFilter(JSON.parse(query.filters));
    }

    const allUniqVariables = Array.from(new Set(allVariables));
    fields =
      (variables && [].concat.apply([], allUniqVariables.map(buildFilter))) ||
      [];

    const filters = (query && query.filters && JSON.parse(query.filters)) || '';

    return (
      <div className='Experiment Review'>
        <div className='header'>
          <ExperimentReviewHeader
            handleGoBackToExplore={this.handleGoBackToExplore}
            handleSaveModel={this.handleSaveModel}
            handleRunAnalysis={this.handleRunAnalysis}
            modelName={apiModel.state.model && apiModel.state.model.title}
            models={apiModel.state.models}
            isMock={apiModel.state.model && apiModel.state.model.isMock}
            handleSelectModel={this.handleSelectModel}
          />
        </div>
        <div className='content'>
          <div className='sidebar'>
            <Model
              model={apiModel.state.model}
              showDatasets={false}
              variables={apiCore.state.variables}
            />
            <Panel className='model'>
              <Panel.Body>
                <Validation
                  isPredictiveMethod={false}
                  datasets={apiCore.state.datasets}
                  query={query}
                  handleUpdateQuery={this.handleUpdateDataset}
                />
              </Panel.Body>
            </Panel>
          </div>
          <div className='results'>
            <Content
              apiMining={apiMining}
              model={apiModel.state.model}
              selectedDatasets={query && query.trainingDatasets}
              lookup={apiCore.lookup}>
              <Panel className='filters' defaultExpanded={false}>
                <Panel.Title toggle={true}>
                  <h3 className={'btn btn-info'}>Filters</h3>
                </Panel.Title>
                <Panel.Collapse>
                  <Panel.Body collapsible={true}>
                    {fields && fields.length > 0 && (
                      <Filter
                        rules={filters}
                        filters={fields}
                        handleChangeFilter={this.handleUpdateFilter}
                      />
                    )}
                  </Panel.Body>
                </Panel.Collapse>
              </Panel>
            </Content>
          </div>
        </div>
      </div>
    );
  }

  private handleUpdateFilter = async (filters: string): Promise<boolean> => {
    const { apiModel, apiMining } = this.props;
    const model = apiModel.state.model;
    if (model) {
      model.query.filters = (filters && JSON.stringify(filters)) || '';
    }
    if (model && !model.isMock) {
      await apiModel.update({ model });
    }
    const query = this.state.query;
    if (query) {
      apiMining.clear();
      this.createMining({ query });
    }

    return Promise.resolve(true);
  };

  private handleSaveModel = async ({ title }: { title: string }) => {
    const { apiModel } = this.props;
    const model = apiModel.state.model;
    const slug = await apiModel.save({ model, title });
    this.setState({ alert: { message: 'Model saved' } });

    const { history } = this.props;
    history.push(`/v3/review/${slug}`);
  };

  private handleRunAnalysis = async () => {
    const { apiModel } = this.props;
    const model = apiModel.state.model;
    await apiModel.update({ model });

    const params = this.urlParams(this.props);
    const slug = params && params.slug;
    const { history } = this.props;
    history.push(`/v3/experiment/${slug}`);
  };

  private handleGoBackToExplore = () => {
    const { apiModel } = this.props;
    const model = apiModel.state.model;
    const query = model && model.query;
    if (query) {
      const variable = query.variables && query.variables.map(v => v.code)[0];
      const covariable =
        (query.coVariables && query.coVariables.map(v => v.code).join(',')) ||
        '';
      const grouping =
        (query.groupings && query.groupings.map(v => v.code).join(',')) || '';
      const trainingDatasets =
        (query.trainingDatasets &&
          query.trainingDatasets.map(v => v.code).join(',')) ||
        '';

      const filterQuery = (query.filters && encodeURI(query.filters)) || '';

      window.location.href = `/explore?configure=true&variable=${variable}&covariable=${covariable}&grouping=${grouping}&filterQuery=${filterQuery}&trainingDatasets=${trainingDatasets}`;
    } else {
      window.location.href = `/explore`;
    }
  };

  private handleSelectModel = (model: MIP.API.IModelResponse) => {
    const { slug } = model;
    const { history } = this.props;
    history.push(`/v3/review/${slug}`);
  };

  private loadModel = async ({ slug }: { slug: string }) => {
    const { apiModel } = this.props;
    await apiModel.one(slug);

    const model = apiModel.state.model;
    if (!model) {
      return this.setState({ alert: { message: 'Fail to load model' } });
    }

    const { query } = model;
    return this.setState({ query });
  };

  private createMining = async ({ query }: { query: MIP.API.IQuery }) => {
    const { apiMining } = this.props;
    const datasets = query.trainingDatasets;

    if (datasets && query) {
      const payload: MIP.API.IMiningPayload = {
        covariables: query.coVariables ? query.coVariables : [],
        datasets,
        filters: query.filters,
        grouping: query.groupings ? query.groupings : [],
        variables: query.variables ? query.variables : []
      };

      await apiMining.allByDataset({ payload });
      return this.setState({
        mining: apiMining.state.minings
      });
    }
  };

  private handleUpdateDataset = (query: MIP.API.IQuery): void => {
    this.setState({ query });
    this.createMining({ query });
  };

  private urlParams = (
    props: IProps
  ):
    | {
        slug: string | undefined;
      }
    | undefined => {
    const { location } = props;
    const slug = location.pathname.split('/').pop() || undefined;

    return { slug: slug !== 'review' ? slug : undefined };
  };
}

export default withRouter(Container);
