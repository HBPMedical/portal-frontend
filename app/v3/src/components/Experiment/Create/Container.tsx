import { APICore, APIExperiment, APIModel } from "@app/components/API";
import Form from "@app/components/Experiment/Create/Form";
import ExperimentCreateHeader from "@app/components/Experiment/Create/Header";
import { Alert, IAlert } from "@app/components/UI/Alert";
import AvailableMethods from "@app/components/UI/AvailableMethods";
import Model from "@app/components/UI/Model";
import Validation from "@app/components/UI/Validation";
import { MIP } from "@app/types";
import * as React from "react";
import { Panel, Tab, Tabs } from "react-bootstrap";
import { RouteComponentProps, withRouter } from "react-router-dom";
import Help from "./Help";

import "../Experiment.css";
interface IProps extends RouteComponentProps<any> {
  apiExperiment: APIExperiment;
  apiCore: APICore;
  apiModel: APIModel;
}

interface IState {
  parameters?: [MIP.API.IMethodPayload];
  query?: MIP.API.IQuery;
  method?: MIP.API.IMethod;
  alert: IAlert;
  kfold: number;
}

class Container extends React.Component<IProps, IState> {
  public state: IState;

  public async componentDidMount() {
    const {
      match: {
        params: { slug }
      }
    } = this.props;
    const { apiModel } = this.props;
    await apiModel.one(slug);

    return this.setState({
      query: apiModel.state.model && apiModel.state.model.query
    });
  }

  public componentWillReceiveProps = (props: any) => {
    const {
      apiModel: {
        state: { model }
      }
    } = props;

    return this.setState({ query: model ? model.query : undefined });
  };

  public render() {
    const { apiCore, apiModel, apiExperiment } = this.props;
    const alert = this.state && this.state.alert;
    const method = this.state && this.state.method;
    const isPredictiveMethod =
      (method && method.type && method.type[0] === "predictive_model") || false;

    return (
      <div className="Experiment">
        <div className="header">
          <ExperimentCreateHeader
            model={apiModel.state.model}
            models={apiModel.state.models}
            experiments={apiExperiment.state.experiments}
            method={this.state && this.state.method}
            handleGoBackToReview={this.handleGoBackToReview}
            handleSelectModel={this.handleSelectModel}
            handleSelectExperiment={this.handleSelectExperiment}
            handleSaveAndRunExperiment={this.handleSaveAndRunExperiment}
          />
        </div>
        <div className="content">
          <div className="sidebar">
            <Model
              model={apiModel.state.model}
              variables={apiCore.state.variables}
            />
          </div>
          <div className="parameters">
            <Panel>
              <Panel.Title>
                <h3>Your Experiment</h3>
              </Panel.Title>
              <Panel.Body>
                {alert && (
                  <Alert
                    message={alert.message}
                    title={alert.title}
                    style={alert.style}
                  />
                )}
                <Tabs
                  defaultActiveKey={1}
                  id="uncontrolled-create-experiment-tab"
                >
                  <Tab eventKey={1} title="Method">
                    <Form
                      method={this.state && this.state.method}
                      parameters={this.state && this.state.parameters}
                      kfold={this.state && this.state.kfold}
                      handleChangeParameters={this.handleChangeParameters}
                      handleChangeKFold={this.handleChangeKFold}
                    >
                      <fieldset style={{ padding: "8px" }}>
                        <Validation
                          isPredictiveMethod={isPredictiveMethod}
                          datasets={apiCore.state.datasets}
                          query={this.state && this.state.query}
                          handleUpdateQuery={this.handleUpdateQuery}
                        />
                      </fieldset>
                    </Form>
                  </Tab>
                  <Tab eventKey={2} title="About running experiments">
                    <Help />
                  </Tab>
                </Tabs>
              </Panel.Body>
            </Panel>
          </div>
          <div className="sidebar2">
            <Panel>
              <Panel.Title>
                <h3>Available Methods</h3>
              </Panel.Title>
              <Panel.Body>
                <AvailableMethods
                  methods={apiCore.state.methods}
                  variables={apiCore.state.variables}
                  handleSelectMethod={this.handleSelectMethod}
                  model={apiModel.state.model}
                />
              </Panel.Body>
            </Panel>
          </div>
        </div>
      </div>
    );
  }
  private handleSelectModel = async (
    model: MIP.API.IModelResponse
  ): Promise<any> => {
    const { slug } = model;
    const { apiModel, history } = this.props;
    history.push(`/v3/experiment/${slug}`);
    if (slug) {
      return await apiModel.one(slug);
    }
  };

  private handleSelectMethod = (method: MIP.API.IMethod): void => {
    const parameters = method && method.parameters;
    this.setState({
      method,
      parameters
    });
  };

  private handleUpdateQuery = (query: MIP.API.IQuery): void => {
    this.setState({ query });
  };

  private handleChangeKFold = (kfold: number) => {
    this.setState({ kfold });
  };

  private handleChangeParameters = (parameters: [MIP.API.IMethodPayload]) => {
    this.setState({ parameters });
  };

  private handleSelectExperiment = async (
    experiment: MIP.API.IExperimentResponse
  ): Promise<any> => {
    const { modelDefinitionId, uuid } = experiment;
    const { apiExperiment, history } = this.props;
    history.push(`/v3/experiment/${modelDefinitionId}/${uuid}`);

    return await apiExperiment.one({ uuid });
  };

  private handleGoBackToReview = () => {
    const { apiModel, history } = this.props;
    const model = apiModel.state.model;
    if (model) {
      history.push(`/v3/review/${model.slug}`);
    }
  };

  private handleSaveAndRunExperiment = async (
    experimentName: string
  ): Promise<any> => {
    const { apiModel, apiExperiment, history } = this.props;

    if (!this.state) {
      this.setState({
        alert: {
          message: "An error occured, please try again later",
          style: "error",
          title: "Info"
        }
      });
      return;
    }
    const { query, parameters } = this.state;

    const model = apiModel.state.model;
    if (!model) {
      this.setState({
        alert: {
          message: "An error occured, please choose a model",
          style: "error",
          title: "Info"
        }
      });
      return;
    }

    const selectedMethod = this.state && this.state.method;
    model.query = query!;
    await apiModel.update({ model });

    const validation =
      model &&
      model.query &&
      model.query.validationDatasets &&
      model.query.validationDatasets.length
        ? true
        : false;

    const validations =
      this.state.kfold > 0
        ? [
            {
              code: "kfold",
              name: "validation",
              parameters: [
                {
                  code: "k",
                  value: this.state.kfold
                }
              ]
            }
          ]
        : [];
    if (!model.slug) {
      this.setState({ alert: { message: "Model was not saved" } });
      return;
    }

    if (!selectedMethod) {
      this.setState({ alert: { message: "Select a method" } });
      return;
    }

    const experiment: MIP.API.IExperimentPayload = {
      algorithms: [
        {
          code: selectedMethod.code,
          name: selectedMethod.code,
          parameters: parameters
            ? parameters.map((p: any) => ({
                ...p,
                value: p.value || p.default_value
              }))
            : [],
          validation
        }
      ],
      model: model.slug,
      name: experimentName,
      validations
    };

    let uuid;
    await apiExperiment.create({ experiment });
    const { experiment: e, error } = apiExperiment.state;

    if (error) {
      this.setState({
        alert: {
          message: `${error}`
        }
      });

      return;
    }

    uuid = e && e.uuid;
    history.push(`/v3/experiment/${model && model.slug}/${uuid}`);
  };
}

export default withRouter(Container);
