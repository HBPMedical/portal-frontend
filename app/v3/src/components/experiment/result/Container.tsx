// tslint:disable:no-console
import { IExperimentResult } from "@app/types";
import * as React from "react";
import { Panel } from "react-bootstrap";
import { RouteComponentProps, withRouter } from "react-router-dom";
import { ExperimentHeader, ExperimentResult, Model } from "../..";
import {
  ExperimentContainer,
  ExperimentListContainer,
  ModelContainer
} from "../../../containers";

import "../Experiment.css";

interface IProps extends RouteComponentProps<any> {
  experimentContainer: ExperimentContainer;
  experimentListContainer: ExperimentListContainer;
  modelContainer: ModelContainer;
}

const methodDisplay = (experiment: IExperimentResult | undefined) => (
  <Panel>
    <Panel.Body>
      <h3>Methods</h3>
      {experiment &&
        experiment.algorithms.map((m: any) => <p key={m.code}>{m.name}</p>)}
      {experiment &&
        experiment.validations &&
        experiment.validations.length > 0 && <h3>Validation</h3>}
      {experiment &&
        experiment.validations &&
        experiment.validations.length > 0 &&
        experiment.validations.map((m: any) => (
          <p key={m.code}>
            {m.code}: {m.parameters.map((p: any) => p.value)}
          </p>
        ))}
    </Panel.Body>
  </Panel>
);

class Experiment extends React.Component<IProps> {
  public async componentDidMount() {
    // Get url parameters
    const { match: matched } = this.props;
    if (!matched) {
      return;
    }
    const { uuid, slug } = matched.params;
    const { experimentContainer, modelContainer } = this.props;

    await modelContainer.load(slug);
    return await experimentContainer.load(uuid);
  }

  public render() {
    const {
      experimentContainer,
      experimentListContainer,
      modelContainer
    } = this.props;
    return (
      <div className="Experiment">
        <div className="header">
          <ExperimentHeader
            experimentContainer={experimentContainer}
            experiments={experimentListContainer.state.experiments}
          />
        </div>

        <div className="sidebar">
          {methodDisplay(experimentContainer.state.experiment)}
          <Model model={modelContainer.state.model} />
        </div>
        <div className="content">
          <ExperimentResult experimentState={experimentContainer.state} />
        </div>
      </div>
    );
  }
}

export default withRouter(Experiment);
