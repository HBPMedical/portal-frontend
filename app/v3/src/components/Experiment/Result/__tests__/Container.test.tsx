import config from "@app/components/API/RequestHeaders";
import Container from "@app/components/Experiment/Result/Container";
import * as React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import * as ReactDOM from "react-dom";
import { APICore, APIExperiment, APIModel } from "@app/components/API";


jest.mock("request-promise-native");
it("renders without crashing", () => {
  const apiExperiment = new APIExperiment(config);
  const apiModel = new APIModel(config);
  const apiCore = new APICore(config);

  const div = document.createElement("div");
  ReactDOM.render(
    <Router>
      <Container
        apiExperiment={apiExperiment}
        apiCore={apiCore}
        apiModel={apiModel}
      />
    </Router>,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});
