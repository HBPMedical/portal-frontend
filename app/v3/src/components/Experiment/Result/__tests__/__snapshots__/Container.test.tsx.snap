// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Result components Result snapshot renders correctly 1`] = `
<div
  className="Experiment"
>
  <div
    className="header"
  >
    <div
      className="panel panel-default"
    >
      <div
        className="panel-body"
      >
        <h3>
          Results of Experiment 
          <strong />
           on
           
          <a
            href="/v3/review/undefined"
            onClick={[Function]}
          />
        </h3>
        <div
          className="actions status"
        >
          <h5
            className="item"
          >
            Created
             
             
            by 
          </h5>
          <div
            className="item text"
          >
            Â 
          </div>
          <div
            className="item"
          >
            <button
              className="btn btn-info"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              SHARE EXPERIMENT
            </button>
          </div>
          <div
            className="item"
          >
            <div
              className="dropdown btn-group btn-group-default"
            >
              <button
                aria-expanded={false}
                aria-haspopup={true}
                className="dropdown-toggle btn btn-default"
                disabled={false}
                id="experiment-dropdown"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                type="button"
              >
                RELATED EXPERIMENTS
                 
                <span
                  className="caret"
                />
              </button>
              <ul
                aria-labelledby="experiment-dropdown"
                className="dropdown-menu"
                role="menu"
              >
                <li
                  className=""
                  role="presentation"
                >
                  <a
                    href="#"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="menuitem"
                    tabIndex="-1"
                  >
                    <strong>
                      Create New Experiment
                    </strong>
                  </a>
                </li>
                <li
                  className=""
                  role="presentation"
                >
                  <a
                    href="#"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="menuitem"
                    tabIndex="-1"
                  >
                    ---------------------------
                  </a>
                </li>
                <div
                  onKeyDown={[Function]}
                  onSelect={[Function]}
                  style={
                    Object {
                      "margin": "8px",
                    }
                  }
                >
                  <p>
                    You have no running experiments.
                  </p>
                  <p>
                    You can start one by selecting a model and configuring an experiment on it.
                  </p>
                </div>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="content"
  >
    <div
      className="sidebar"
    >
      <div
        className="model panel panel-default"
      >
        <div
          className="panel-title"
        >
          <h3>
            Model 
            <strong />
          </h3>
        </div>
        <div
          className="panel-body"
        />
      </div>
    </div>
    <div
      className="results"
    >
      <div
        className="panel panel-default"
      >
        <div
          className="panel-title"
        >
          <h3>
            Your Experiment
          </h3>
        </div>
        <div
          className="panel-body"
        >
          <div
            className="loading"
          >
            <h3>
              Your experiment is currently running...
            </h3>
            <p>
              Please check back in a few minutes. This page will automatically refresh once your experiment has finished executing.
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="sidebar2"
    />
  </div>
</div>
`;
